zproject/backends.py

```diff
@@ -866,6 +884,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm.id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names)
+
+            member_group_names = [
+                membership.user_group.name
+                for membership in UserGroupMembership.objects.filter(user_profile=user_profile)
+            ]
+
+            ldap_logger.debug("user groups: %s", repr(member_group_names))
+
+            for group_name in ldap_group_name_set:
+                if group_name not in set(settings.LDAP_GROUP_SYNC_LIST):
+                    continue  # this group shall not be synced!
+
+                ldap_logger.debug("group: %s", str(group_name))
+
+                group = UserGroup.objects.get(name=group_name, realm=user_profile.realm)
+
+                # we continue above if group should not be synced and create group if not exists no need to check the group again
+                if not is_user_in_group(
+                    user_group=group, user=user_profile, direct_member_only=True
+                ):
+                    ldap_logger.debug("add %s to %s", user_profile.full_name, str(group_name))
+                    UserGroupMembership.objects.create(
+                        user_group=group,
+                        user_profile=user_profile,
+                    )
```

Comment:
Also should be more efficient to just build a list of `UserGroupMembership`s here without doing db queries with `.create()` - and instead `.bulk_create()` the whole list after the loop. That'll be just a single query.